promql/promqltest/test.go

```diff
@@ -693,29 +792,79 @@ func (ev *evalCmd) expectMetric(pos int, m labels.Labels, vals ...parser.Sequenc
 	ev.expected[h] = entry{pos: pos, vals: vals}
 }
 
+// validateExpectedAnnotations validates expected messages and regex match actual annotations.
+func validateExpectedAnnotations(expr string, expectedAnnotations []expectCmd, actualAnnotations []string, line int, annotationType string) error {
+	if len(expectedAnnotations) == 0 {
+		return nil
+	}
+
+	if len(actualAnnotations) == 0 {
+		return fmt.Errorf("expected %s annotations but none were found for query %q (line %d)", annotationType, expr, line)
+	}
+
+	// Check if all expected annotations are found in actual.
+	for _, e := range expectedAnnotations {
+		matchFound := slices.ContainsFunc(actualAnnotations, e.CheckMatch)
+		if !matchFound {
+			return fmt.Errorf("expected %s annotation %q but no matching annotation was found for query %q (line %d)", annotationType, e.String(), expr, line)
+		}
+	}
+
+	// Check if all actual annotations have a corresponding expected annotation.
+	for _, anno := range actualAnnotations {
+		matchFound := slices.ContainsFunc(expectedAnnotations, func(e expectCmd) bool {
+			return e.CheckMatch(anno)
+		})
+		if !matchFound {
+			return fmt.Errorf("unexpected %s annotation %q found for query %s (line %d)", annotationType, anno, expr, line)
+		}
+	}
+
+	return nil
+}
+
 // checkAnnotations asserts if the annotations match the expectations.
 func (ev *evalCmd) checkAnnotations(expr string, annos annotations.Annotations) error {
 	countWarnings, countInfo := annos.CountWarningsAndInfo()
 	switch {
 	case ev.ordered:
 		// Ignore annotations if testing for order.
```

Comment:
This case doesn't do anything anymore. It can also go. (It only existed previously to avoid the check for no annotations in this case. But we don't do that check anymore in this switch statement.)